### DISCUSSION TITLE
No Spotify login?

### DISCUSSION THREAD

**Discussion by rickballard** [micro-nova/AmpliPi#315] [Q&A]
The web app appears to need a Spotify device on the LAN, before Spotify can be added to the list of available inputs. I have a Spotify account and I had expected to be able to just enter those login details into the web app.  Am I missing something?

**Comment by Lohrer** [comment]
AmpliPi uses [Spotify Connect](https://support.spotify.com/us/article/spotify-connect/), which means it acts as a selectable device to play from. You control what's playing from the Spotify phone app, website, or desktop client. To select AmpliPi as the playback device, you have to create a Spotify stream if it doesn't already exist (the Stream Name will be what shows up in your Spotify app):
![Spotify Stream](https://user-images.githubusercontent.com/6646216/163583589-f9d195fd-20e1-4c49-9639-f566adb1a1d3.png)

Then select that stream as an input to a source:
![Select Spotify](https://user-images.githubusercontent.com/6646216/163585079-5c18a943-40c8-4a09-918f-0dcc384fbb99.png)

Finally open your Spotify client of choice and click the devices button ![Devices](https://cdn.sanity.io/images/tsbk0zvv/production/872039fca8973409cb01bbce6eb0fb42b5ddace2-30x30.png?h=30) and select "AmpliPi" or whatever you named your Spotify stream. The Spotify Connect link above shows more info on this step as well. Once you play a song you should see the info come up in AmpliPi:
![Play to AmpliPi](https://user-images.githubusercontent.com/6646216/163584572-39b13fa0-658f-4a55-bc67-97f3737cc7d6.png)

Also see discussion #210 about setting up Spotify, but I believe I've covered everything here.

**Comment by rickballard** [comment]
Thank you for your detailed and excellent explanation.  Danged if I can find an AmpliPi device on my smartphone's Spotify app, but a lot of other devices are listed, like my old HiFi, some smartTV's and Google speakers, but no AmpliPi. 

Anyway, to demonstrate a hyperthetical use case, in the event that a Spotify Connect feature ever shows up for me, here is a demonstration of my cringe-worthy newbie-ness, (this has got to be wrong somewhere):

1. First I launch my Spotify app on my smartphone and set it to play a song...
2. but only after a secret Spotify Connect feature has been turned on and named "Spanky" in my smartphone's Spotify app menu...
3. Spanky is the name of the secret audio-stream that is playing the afore-mentioned song to my LAN, WiFi and guest-WiFi...
4. then I launch my smartphone's AmpliPi app and point it at the secretive "Spanky" option, if it shows up,
5. then the audio volume-controls may need to be adjusted independently, in both my Spotify app and my AmpliPi app...
6. then when I turn the music off and go to bed, I must remember to turn it off in both apps...
7. then, when I try to listen to music the next morning, my AmpliPi will remember Spanky, but my Spotify app will not.
8. I tell the wife about all of this.
9. the wife kills me.
10. I find peace at last.

The Spotify website does not have a Spotify Connect feature that allows an audio-stream to run on my always-on server or some such, playing songs permanently in the LAN's background.  So, I assume there is no chance we will ever be able to develop a single-press solution, even if we buy the future (wife-friendly) ethernet-keypads...?

**Reply by Lohrer** [reply]
You have the 'device' setup backwards, if you want AmpliPi to show up as the device "Spanky" you set that up in the AmpliPi app first, then open Spotify on your smartphone.

1. Launch your smartphone's AmpliPi app (or the web app on a browser)
   1. Click the gear icon tab (settings)
   1. Click "Add New Stream"
   1. Set "Stream Type" to "Spotify Device"
   1. Set "Stream Name" to "Spanky"
   1. This step only has to be done ONCE, now you have a "Spanky" spotify 'device' to play to. This step corresponds to the first picture in my post above.
1. Still in the AmpliPi app, select the source 1 tab (or 2, or 3, or 4)
   1. Set the dropdown menu to "Spanky - spotify" - this will start up the "Spanky" spotify 'device' and have it play to whichever zones you add to that source in AmpliPi.
   1. This corresponds to the second picture in my post above.
   1. This step can be done ONCE if you don't ever care to change the source/zone setup in AmpliPi.
   1. If you DO ever de-select the "Spanky - spotify" input on the source, it will also disappear on your Spotify app.
1. Open your smartphone or desktop's Spotify app, click the devices button and select "Spanky". Start playing a song!

So your steps become:
4\. ~~I launch my smartphone's AmpliPi app and point it at the "Spanky" option, if it shows up~~ I create the "Spanky" Spotify stream in my smartphone's AmpliPi app, and set a source to play it.
1. I launch my Spotify app on my smartphone [and set it to play a song] - playing can come after connecting or before, doesn't matter.
2. ~~but only after a secret Spotify Connect feature has been turned on and named "Spanky" in my smartphone's Spotify app menu...~~ No special Spotify settings required, just press the devices button in your Spotify app!
3. ~~Spanky is the name of the secret audio-stream that is playing the afore-mentioned song to my LAN, WiFi and guest-WiFi...~~ I'm not sure what you mean by this step, Spotify doesn't "stream to your LAN", it plays directly from Spotify's servers. AmpliPi needs access to the internet for Spotify to work. The first time you connect from your smartphone Spotify app you may have to be on the same LAN as the AmpliPi but don't quote me on that, I know it works across the internet for my connected devices.
5. then the audio volume-controls may need to be adjusted independently, in both my Spotify app and my AmpliPi app... Currently Spotify can only reduce the volume, but this bug is tracked in #194 and is in the queue to be fixed.
6. then when I turn the music off and go to bed, I must remember to turn it off in ~~both~~ either app...Just pause in either, or mute in either, or disconnect in either...there are got options!
7. then, when I try to listen to music the next morning, my AmpliPi will remember Spanky, ~~but my Spotify app will not.~~
8. I tell the wife about all of this.
9. the wife ~~kills~~ loves me.
10. I find peace at last.

I personally use Spotify the most of all of the currently available AmpliPi stream options. There ARE a couple outstanding bugs right now (see #271) but they are next up in my queue of things to fix. Still doesn't stop me from playing Spotify on AmpliPi all day though! And the only time I have to touch the AmpliPi app at @gorski123's house is when someone steals my zone to play something else.

**Reply by rickballard** [reply]
You are now my favourite AmpliPi person!  Thank you, thank you, thank you!

**Reply by rickballard** [reply]
Still get that hissing whistle in the background though, when listening to no input or the Spotify stream.  I wonder if it is due to that antennae thumb-drive thingy....
